### DISCUSSION TITLE
correct/recommended setup of changes API

### DISCUSSION THREAD

**Discussion by wallaceturner** [ravendb/ravendb#16694] [General]
We have come across an issue recently where restarting ravendb broke the changes API (v 5.4.1) - what we observed is that `IConnectableChanges.Connected` returned false and continued to return false even after ravendb  restarted.

We were experimenting and then suddenly things just started to work again - that is the changes API reconnected after a ravendb reboot without us changing any of the underlying code.

I refer to [your changes api docs](https://ravendb.net/docs/article-page/5.4/csharp/client-api/changes/what-is-changes-api)

Do you **need** to call `EnsureConnectedNow` ? same question also for `EnsureSubscribedNow`  ?

there is this [stackoverflow post](https://stackoverflow.com/questions/64722546/ravendb-changes-api-never-fires)  which differs greatly from the offical docs

**Comment by garayx** [comment]
Hi
`EnsureSubscribedNow()` is good to ensure that Changes API is connected and subscribed by the moment this method returns.
In regular usage, you actually do not care about calling `EnsureSubscribedNow()`, due to how Changes API works, there is no delivery guaranteed, it might get disconnected and some events might be skipped, for example on server restart.

**Reply by ayende** [reply]
Note that it should attempt to reconnect on failure, see:

https://github.com/ravendb/ravendb/blob/04444b6b4cab38cdf3033323160a3c8faf9ad22f/src/Raven.Client/Documents/Changes/AbstractDatabaseChanges.cs#L299

You can subscribe to `ConnectionStatusChanged` to get changes.

On failure, it should wait 1 sec and reconenct

https://github.com/ravendb/ravendb/blob/04444b6b4cab38cdf3033323160a3c8faf9ad22f/src/Raven.Client/Documents/Changes/AbstractDatabaseChanges.cs#L378

**Comment by wallaceturner** [comment]
>> it might get disconnected and some events might be skipped, for example on server restart.
we arent concerned with that (we have a way of handling this internally)


This appears to be a random/non-determinstic issue - the 'stable' recreate was
- subscribe
- restart server
- insert doc that is subscribed in step 1)
- observe that no `OnNext` is called

nb we were able to workaround the issue with the following tickle:
```
store.Changes().ConnectionStatusChanged += DataImporterService_ConnectionStatusChanged;

private void DataImporterService_ConnectionStatusChanged(object sender, EventArgs e)
        {
            var changes = (IDatabaseChanges)sender;
            if (!changes.Connected)
            {
                changes.EnsureConnectedNow();
            }
            
        }
```

**Reply by ayende** [reply]
Note that it may very well lose items while it is attempting to reconnect.
That is explicitly part of the contract, if you need reliability, you should use a subscription.
Changes is for "mostly there" and fast reactions

**Reply by wallaceturner** [reply]
@ayende  as per [the docs](https://ravendb.net/docs/article-page/5.4/csharp/client-api/data-subscriptions/what-are-data-subscriptions)
>The client then processes the batch and will receive the next one only after it acknowledges the batch was processed.

how does the client "acknowledge the batch was processed" ?

there is no obvious method in `SubscriptionBatch` - is the acknowledgement simply returning from `SubscriptionWorker.Run` without an Exception?

**Reply by ppekrol** [reply]
That is correct, it is done automatically if there is no exception during batch processing.

**Comment by wallaceturner** [comment]
dupe of https://github.com/ravendb/ravendb/discussions/17474
